Behavior difference: declaration emit synthesizes extensionless `import()` specifier under `allowImportingTsExtensions`, invalid in nodenext

When a project uses `allowImportingTsExtensions`-style relative imports (`import { x } from './helper.ts'`), tsgo's declaration emit prints synthesized `import()` type references to package subpaths **without their file extension**. Under `--module nodenext` the extensionless specifier does not resolve, so the emitted `.d.ts` is invalid. Emit succeeds with no diagnostic.

Reproduced with `typescript@7.0.2` and nightly `@typescript/native-preview@7.0.0-dev.20260707.2`.

## Steps to reproduce

Self-contained — no `npm install`; the dependency is materialized by the script (a plain CJS package with no `exports` map, imported via an explicit-extension subpath, like fp-ts among others):

```bash
mkdir repro && cd repro && mkdir -p src node_modules/mylib/lib

cat > package.json <<'EOF'
{ "name": "repro", "type": "module" }
EOF

cat > tsconfig.json <<'EOF'
{
  "compilerOptions": {
    "module": "nodenext",
    "declaration": true,
    "emitDeclarationOnly": true,
    "allowImportingTsExtensions": true,
    "rootDir": "src",
    "outDir": "out"
  },
  "include": ["src"]
}
EOF

cat > node_modules/mylib/package.json <<'EOF'
{ "name": "mylib", "version": "1.0.0", "main": "./lib/index.js" }
EOF

cat > node_modules/mylib/lib/Box.js <<'EOF'
exports.box = (value) => ({ value });
EOF

cat > node_modules/mylib/lib/Box.d.ts <<'EOF'
export interface Box<T> { readonly value: T }
export declare function box<T>(value: T): Box<T>;
EOF

cat > src/helper.ts <<'EOF'
import * as B from 'mylib/lib/Box.js';
export const inner = B.box('hello');
EOF

cat > src/mod.ts <<'EOF'
import { inner } from './helper.ts';
export const out = inner;
EOF
```

Then compile with each compiler and inspect `out/mod.d.ts`, as shown in the two sections below.

## Behavior with `typescript@6.0`

```bash
npx -y -p typescript@6.0.2 tsc -p tsconfig.json
cat out/mod.d.ts
```

Emits a resolvable specifier:

```ts
export declare const out: import("mylib/lib/Box.js").Box<string>;
```

## Behavior with `tsgo`

```bash
npx -y -p typescript@7.0.2 tsc -p tsconfig.json
cat out/mod.d.ts
```

The extension is dropped:

```ts
export declare const out: import("mylib/lib/Box").Box<string>;
```

## The trigger is the `.ts`-extension relative import

One-character bisect: change `src/mod.ts` to import `'./helper.js'` instead of `'./helper.ts'` and tsgo emits the extension correctly. Generics, type aliases, and `rewriteRelativeImportExtensions` were all eliminated as factors.

## The emitted file is invalid

```bash
cat > tsconfig.verify.json <<'EOF'
{
  "compilerOptions": { "module": "nodenext", "noEmit": true, "skipLibCheck": false },
  "include": ["out/mod.d.ts"]
}
EOF
npx -y -p typescript@7.0.2 tsc -p tsconfig.verify.json
```

```
out/mod.d.ts(1,34): error TS2307: Cannot find module 'mylib/lib/Box' or its corresponding type declarations.
```

(`typescript@6.0.2` rejects it identically)

In a monorepo with `skipLibCheck: true` this is silent — the broken reference degrades to an error type and surfaces as unrelated-looking inference failures (`'x' is of type 'unknown'`) in downstream packages.

---

*Disclosure: this bug was found, minimized, and verified interactively by a human (me) working with Claude Code; the write-up was drafted by the model and reviewed by me. All commands and outputs above were executed and confirmed against the stated versions.*

--- a/internal/modulespecifiers/preferences.go
+++ b/internal/modulespecifiers/preferences.go
@@ -174,7 +174,13 @@ func GetAllowedEndingsInPreferredOrder(
 	moduleResolution := compilerOptions.GetModuleResolutionKind()
 	moduleResolutionIsNodeNext := core.ModuleResolutionKindNode16 <= moduleResolution && moduleResolution <= core.ModuleResolutionKindNodeNext
 	allowImportingTsExtension := shouldAllowImportingTsExtension(compilerOptions, importingSourceFile.FileName())
-	if syntaxImpliedNodeFormat == core.ResolutionModeESM && moduleResolutionIsNodeNext {
+	// TypeScript uses `(syntaxImpliedNodeFormat ?? impliedNodeFormat)` here - fall back to the
+	// file's default resolution mode when no syntax-implied mode is given.
+	effectiveSyntaxMode := syntaxImpliedNodeFormat
+	if effectiveSyntaxMode == core.ResolutionModeNone {
+		effectiveSyntaxMode = resolutionMode
+	}
+	if effectiveSyntaxMode == core.ResolutionModeESM && moduleResolutionIsNodeNext {
 		if allowImportingTsExtension {
 			return []ModuleSpecifierEnding{ModuleSpecifierEndingTsExtension, ModuleSpecifierEndingJsExtension}
 		}

--- a/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.js
+++ b/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.js
@@ -1,0 +1,28 @@
+//// [tests/cases/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.ts] ////
+
+//// [package.json]
+{ "type": "module" }
+
+//// [package.json]
+{ "name": "mylib", "version": "1.0.0", "main": "./lib/index.js" }
+
+//// [Box.d.ts]
+export interface Box<T> { readonly value: T }
+export declare function box<T>(value: T): Box<T>;
+
+//// [helper.ts]
+import * as B from 'mylib/lib/Box.js';
+export const inner = B.box('hello');
+
+//// [mod.ts]
+import { inner } from './helper.ts';
+export const out = inner;
+
+
+
+
+//// [helper.d.ts]
+import * as B from 'mylib/lib/Box.js';
+export declare const inner: B.Box<string>;
+//// [mod.d.ts]
+export declare const out: import("mylib/lib/Box.js").Box<string>;

--- a/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.symbols
+++ b/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.symbols
@@ -1,0 +1,35 @@
+//// [tests/cases/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.ts] ////
+
+=== /node_modules/mylib/lib/Box.d.ts ===
+export interface Box<T> { readonly value: T }
+>Box : Symbol(Box, Decl(Box.d.ts, 0, 0))
+>T : Symbol(T, Decl(Box.d.ts, 0, 21))
+>value : Symbol(Box.value, Decl(Box.d.ts, 0, 25))
+>T : Symbol(T, Decl(Box.d.ts, 0, 21))
+
+export declare function box<T>(value: T): Box<T>;
+>box : Symbol(box, Decl(Box.d.ts, 0, 45))
+>T : Symbol(T, Decl(Box.d.ts, 1, 28))
+>value : Symbol(value, Decl(Box.d.ts, 1, 31))
+>T : Symbol(T, Decl(Box.d.ts, 1, 28))
+>Box : Symbol(Box, Decl(Box.d.ts, 0, 0))
+>T : Symbol(T, Decl(Box.d.ts, 1, 28))
+
+=== /helper.ts ===
+import * as B from 'mylib/lib/Box.js';
+>B : Symbol(B, Decl(helper.ts, 0, 6))
+
+export const inner = B.box('hello');
+>inner : Symbol(inner, Decl(helper.ts, 1, 12))
+>B.box : Symbol(B.box, Decl(Box.d.ts, 0, 45))
+>B : Symbol(B, Decl(helper.ts, 0, 6))
+>box : Symbol(B.box, Decl(Box.d.ts, 0, 45))
+
+=== /mod.ts ===
+import { inner } from './helper.ts';
+>inner : Symbol(inner, Decl(mod.ts, 0, 8))
+
+export const out = inner;
+>out : Symbol(out, Decl(mod.ts, 1, 12))
+>inner : Symbol(inner, Decl(mod.ts, 0, 8))
+

--- a/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.types
+++ b/testdata/baselines/reference/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.types
@@ -1,0 +1,30 @@
+//// [tests/cases/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.ts] ////
+
+=== /node_modules/mylib/lib/Box.d.ts ===
+export interface Box<T> { readonly value: T }
+>value : T
+
+export declare function box<T>(value: T): Box<T>;
+>box : <T>(value: T) => Box<T>
+>value : T
+
+=== /helper.ts ===
+import * as B from 'mylib/lib/Box.js';
+>B : typeof B
+
+export const inner = B.box('hello');
+>inner : B.Box<string>
+>B.box('hello') : B.Box<string>
+>B.box : <T>(value: T) => B.Box<T>
+>B : typeof B
+>box : <T>(value: T) => B.Box<T>
+>'hello' : "hello"
+
+=== /mod.ts ===
+import { inner } from './helper.ts';
+>inner : import("mylib/lib/Box.js").Box<string>
+
+export const out = inner;
+>out : import("mylib/lib/Box.js").Box<string>
+>inner : import("mylib/lib/Box.js").Box<string>
+

--- a/testdata/tests/cases/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.ts
+++ b/testdata/tests/cases/compiler/declarationEmitImportSpecifierWithTsExtensionRelativeImport.ts
@@ -1,0 +1,26 @@
+// @module: nodenext
+// @declaration: true
+// @emitDeclarationOnly: true
+// @allowImportingTsExtensions: true
+
+// Regression test for https://github.com/microsoft/typescript-go/issues/4616
+// When a file uses `.ts` extension relative imports with allowImportingTsExtensions,
+// declaration emit should preserve `.js` extensions on non-relative (node_modules) imports.
+
+// @Filename: /package.json
+{ "type": "module" }
+
+// @Filename: /node_modules/mylib/package.json
+{ "name": "mylib", "version": "1.0.0", "main": "./lib/index.js" }
+
+// @Filename: /node_modules/mylib/lib/Box.d.ts
+export interface Box<T> { readonly value: T }
+export declare function box<T>(value: T): Box<T>;
+
+// @Filename: /helper.ts
+import * as B from 'mylib/lib/Box.js';
+export const inner = B.box('hello');
+
+// @Filename: /mod.ts
+import { inner } from './helper.ts';
+export const out = inner;

